internal/mongodoc/doc.go

```diff
@@ -181,8 +194,20 @@ type BaseEntity struct {
 	CommonInfo map[string][]byte `bson:",omitempty" json:",omitempty"`
 }
 
-func (e *BaseEntity) CanonicalURL() *charm.URL {
-	return e.URL
+// CanonicalURLs returns the canonical URLs that can be used
+// to look up the entity. For non-promulgated entities,
+// there will only be one element in the result.
+func (e *BaseEntity) CanonicalURLs() []*charm.URL {
+	// Larger capacity than strictly necessary to allow
+	// a caller to append to the returned slice without allocating.
+	urls := make([]*charm.URL, 1, 3)
+	urls[0] = copyURL(e.URL)
+	if e.Promulgated {
```

Comment:
There should be no such thing as a promulgated base URL, having the concept really worries me about the fragility of the system. An entity has a single base URL that base URL always has an owner. It might be a property of a base entity that new entities are promulgated, but I don't consider that that reasonably makes the base entity itself promulgated. Attempting to address a base entity by a promulgated name seems like it will lead to disaster.